File: rag/src/scanner.py
"""
scanner.py - 知识源文件扫描器。

扫描本地知识包、skill 规则/参考文档和项目模板，
输出相对路径与标准化 source_type。
"""

import re
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent

INCLUDE_PATTERNS = [
    ".harness/knowledge/builtin/**/*.md",
    ".harness/knowledge/remote/**/*.md",
    ".harness/skills/**/references/*.md",
    ".harness/skills/**/rules/*.md",
    ".harness/project-templates/*.md",
]

EXCLUDE_PATTERNS = [
    "*planning*",
    "*legacy-skills*",
    ".harness/knowledge/user/*",
    "*agents*",
    "*/current-project/*",
    "*/README*",
    "*memory*",
    "*cases*",
    ".harness/rules/*",
]


def _glob_to_regex(pattern):
    """把本项目使用的简单 glob 模式转换成正则。"""
    parts = pattern.split("/")
    regex_parts = []
    for part in parts:
        if part == "**":
            regex_parts.append("(.+/)?")
        elif part == "*":
            regex_parts.append("[^/]*")
        else:
            escaped = re.escape(part).replace(r"\*", "[^/]*").replace(r"\?", ".")
            regex_parts.append(escaped)
    return "^" + "/".join(regex_parts) + "$"


def _path_matches(path, pattern_list):
    path_str = str(path).replace("\\", "/")
    for pattern in pattern_list:
        if re.match(_glob_to_regex(pattern), path_str):
            return True
    return False


def _infer_source_type(pattern):
    if "/knowledge/" in pattern:
        return "knowledge"
    if "/rules/" in pattern:
        return "rule"
    if "/references/" in pattern:
        return "reference"
    if "/project-templates/" in pattern:
        return "project"
    return "reference"


def scan_knowledge_files():
    """扫描所有可索引的知识文件。"""
    results = []

    for pattern in INCLUDE_PATTERNS:
        source_type = _infer_source_type(pattern)
        matched_files = [p for p in PROJECT_ROOT.glob(pattern) if p.is_file()]

        for file_path in matched_files:
            rel_path = str(file_path.relative_to(PROJECT_ROOT)).replace("\\", "/")
            if _path_matches(rel_path, EXCLUDE_PATTERNS):
                continue

            results.append({
                "rel_path": rel_path,
                "source_type": source_type,
            })

    seen = set()
    unique = []
    for item in results:
        if item["rel_path"] not in seen:
            seen.add(item["rel_path"])
            unique.append(item)

    unique.sort(key=lambda x: x["rel_path"])
    print(f"[scanner] 扫描到 {len(unique)} 个知识文件")
    return unique

File: rag/src/scanner.js
/**
 * scanner.js — 知识源扫描器
 *
 * 职责：
 * - 按 sources.json 中的 include/exclude 模式扫描文件系统
 * - 返回匹配的知识文件列表
 */

import { readFileSync, readdirSync, statSync } from 'fs';
import { join, resolve, dirname, basename } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = resolve(__dirname, '../..');

/**
 * 加载 sources.json 配置
 */
export function loadSourceConfig() {
  const configPath = join(PROJECT_ROOT, 'rag/config/sources.json');
  return JSON.parse(readFileSync(configPath, 'utf-8'));
}

/**
 * 将 glob 模式转为正则
 */
function globToRegex(pattern) {
  const normalized = pattern.replace(/\\/g, '/');
  let regexStr = normalized
    .replace(/\./g, '\\.')
    .replace(/\*\*/g, '___DBL___')
    .replace(/\*/g, '[^/]*')
    .replace(/___DBL___/g, '.*');
  return new RegExp(`^${regexStr}$`);
}

/**
 * 递归收集文件
 */
function collectFiles(dir, includePattern, excludePatterns) {
  const results = [];
  let entries;
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return results;
  }

  for (const entry of entries) {
    const fullPath = join(dir, entry.name);
    const relPath = fullPath.replace(/\\/g, '/');

    // 排除 .git, node_modules 等
    if (entry.name.startsWith('.') && entry.name !== '.harness') continue;
    if (entry.name === 'node_modules') continue;

    if (entry.isDirectory()) {
      results.push(...collectFiles(fullPath, includePattern, excludePatterns));
    } else if (entry.isFile() && entry.name.endsWith('.md')) {
      // 检查是否匹配 include 模式
      if (!includePattern.test(relPath)) continue;

      // 检查是否在 exclude 中
      let excluded = false;
      for (const ex of excludePatterns) {
        if (ex.test(relPath)) { excluded = true; break; }
      }
      if (excluded) continue;

      results.push(relPath.replace(PROJECT_ROOT.replace(/\\/g, '/') + '/', ''));
    }
  }

  return results;
}

/**
 * 扫描知识文件
 * @returns {Array<{path: string, source_type: string}>}
 */
export function scanKnowledgeFiles() {
  const config = loadSourceConfig();

  const includePatterns = {};
  for (const [type, pattern] of Object.entries(config.source_types)) {
    includePatterns[type] = globToRegex(join(PROJECT_ROOT, pattern).replace(/\\/g, '/'));
  }

  const excludeRegexes = config.exclude_patterns.map(p =>
    globToRegex(p.startsWith('**') ? join(PROJECT_ROOT, p).replace(/\\/g, '/') : join(PROJECT_ROOT, p).replace(/\\/g, '/'))
  );

  const allFiles = [];
  const seen = new Set();

  for (const [sourceType, includeRegex] of Object.entries(includePatterns)) {
    const matches = collectFiles(PROJECT_ROOT, includeRegex, excludeRegexes);
    for (const relPath of matches) {
      if (seen.has(relPath)) continue;
      seen.add(relPath);

      allFiles.push({
        path: relPath,
        source_type: sourceType
      });
    }
  }

  return allFiles;
}
